File: src/db/queries/apiStats.ts
//! External-API (yfinance) rate-limit accounting. Ports OLD/src/db/api_stats.rs.
//! Uses the raw better-sqlite3 handle so the atomic check-and-increment UPDATE
//! (julianday interval guard) is a verbatim translation of the original SQL.
import { sqliteClient as sdb } from "@/db";
import { jstNow, jstToday } from "@/lib/format";

const MIN_INTERVAL_MINUTES = 96; // 1440 / 15 requests per day
const MAX_REQUESTS_PER_DAY = 15;

/** `YYYY-MM-DD HH:MM:SS` in JST, matching the stored format. */
function fmtJstTs(d: Date): string {
  return d.toISOString().slice(0, 19).replace("T", " ");
}

function ensureStatsForToday(): void {
  sdb
    .prepare(
      `INSERT OR IGNORE INTO api_request_stats (reset_date, last_request_time, request_count) VALUES (?, NULL, 0)`,
    )
    .run(jstToday());
}

export interface ApiRequestStats {
  lastRequestTime: string | null;
  requestCount: number;
  resetDate: string;
}

export function getStats(): ApiRequestStats {
  ensureStatsForToday();
  const today = jstToday();
  const row = sdb
    .prepare(
      `SELECT last_request_time AS lastRequestTime, request_count AS requestCount, reset_date AS resetDate
       FROM api_request_stats WHERE reset_date = ? LIMIT 1`,
    )
    .get(today) as ApiRequestStats | undefined;
  return row ?? { lastRequestTime: null, requestCount: 0, resetDate: today };
}

/** Informational check (use tryRecordApiRequest to actually claim a slot). */
export function canRequestApi(): {
  canRequest: boolean;
  minutesUntilNext: number;
} {
  const stats = getStats();
  if (stats.requestCount >= MAX_REQUESTS_PER_DAY) {
    return { canRequest: false, minutesUntilNext: 0 };
  }
  if (!stats.lastRequestTime) {
    return { canRequest: true, minutesUntilNext: 0 };
  }
  // Stored timestamp is JST wall-clock; parse as UTC so the delta vs jstNow()
  // (also UTC-encoded JST) is the real elapsed time.
  const last = new Date(stats.lastRequestTime.replace(" ", "T") + "Z").getTime();
  const elapsedMin = Math.floor((jstNow().getTime() - last) / 60000);
  const remaining = MIN_INTERVAL_MINUTES - elapsedMin;
  return remaining <= 0
    ? { canRequest: true, minutesUntilNext: 0 }
    : { canRequest: false, minutesUntilNext: remaining };
}

/**
 * Atomically claim a request slot. The WHERE clause enforces both the daily
 * cap and the minimum interval, so concurrent callers can't both succeed.
 */
export function tryRecordApiRequest(): boolean {
  ensureStatsForToday();
  const today = jstToday();
  const nowStr = fmtJstTs(jstNow());
  const res = sdb
    .prepare(
      `UPDATE api_request_stats
         SET last_request_time = ?, request_count = request_count + 1
       WHERE reset_date = ?
         AND request_count < ?
         AND (last_request_time IS NULL
              OR (julianday(?) - julianday(last_request_time)) * 1440 >= ?)`,
    )
    .run(nowStr, today, MAX_REQUESTS_PER_DAY, nowStr, MIN_INTERVAL_MINUTES);
  return res.changes > 0;
}

/** Force-record (daily scheduler / owner-initiated) — bypasses the limiter. */
export function recordApiRequestForced(): void {
  ensureStatsForToday();
  const today = jstToday();
  const nowStr = fmtJstTs(jstNow());
  sdb
    .prepare(
      `UPDATE api_request_stats
         SET last_request_time = ?, request_count = request_count + 1
       WHERE reset_date = ?`,
    )
    .run(nowStr, today);
}

export function requestsRemaining(): number {
  return Math.max(MAX_REQUESTS_PER_DAY - getStats().requestCount, 0);
}

export { MAX_REQUESTS_PER_DAY, MIN_INTERVAL_MINUTES };

File: src/db/api_stats.rs
use chrono::{DateTime, FixedOffset, NaiveDate, NaiveDateTime, TimeZone};
use sqlx::SqlitePool;

use crate::util::{jst, jst_today};

const MIN_INTERVAL_MINUTES: i64 = 96; // 1440 / 15 requests per day
const MAX_REQUESTS_PER_DAY: i64 = 15;

fn jst_now() -> DateTime<FixedOffset> {
    chrono::Utc::now().with_timezone(&jst())
}

#[derive(Debug, Clone)]
pub struct ApiRequestStats {
    pub last_request_time: Option<DateTime<FixedOffset>>,
    pub request_count: i64,
    pub reset_date: NaiveDate,
}

/// Initialize stats row for today if it doesn't already exist.
async fn ensure_stats_for_today(pool: &SqlitePool) -> sqlx::Result<()> {
    let today = jst_today();

    sqlx::query(
        r#"
        INSERT OR IGNORE INTO api_request_stats (reset_date, last_request_time, request_count)
        VALUES (?, NULL, 0)
        "#,
    )
    .bind(today)
    .execute(pool)
    .await?;

    Ok(())
}

/// Get current API request stats for today.
pub async fn get_stats(pool: &SqlitePool) -> sqlx::Result<ApiRequestStats> {
    ensure_stats_for_today(pool).await?;

    let today = jst_today();

    let row: (Option<String>, i64, String) = sqlx::query_as(
        r#"
        SELECT last_request_time, request_count, reset_date
        FROM api_request_stats
        WHERE reset_date = ?
        LIMIT 1
        "#,
    )
    .bind(today)
    .fetch_one(pool)
    .await?;

    let last_request_time = row.0.and_then(|ts| {
        NaiveDateTime::parse_from_str(&ts, "%Y-%m-%d %H:%M:%S")
            .ok()
            .and_then(|naive| jst().from_local_datetime(&naive).single())
    });

    Ok(ApiRequestStats {
        last_request_time,
        request_count: row.1,
        reset_date: NaiveDate::parse_from_str(&row.2, "%Y-%m-%d").unwrap_or(today),
    })
}

/// Speculative check: is an external API request allowed right now?
/// Returns (can_request, minutes_until_next_allowed).
///
/// **NOTE**: this is informational only. To actually claim a slot, call
/// [`try_record_api_request`] which performs the same check atomically.
pub async fn can_request_api(pool: &SqlitePool) -> sqlx::Result<(bool, i64)> {
    let stats = get_stats(pool).await?;

    if stats.request_count >= MAX_REQUESTS_PER_DAY {
        return Ok((false, 0));
    }

    let Some(last) = stats.last_request_time else {
        return Ok((true, 0));
    };

    let now = jst_now();
    let elapsed_minutes = now.signed_duration_since(last).num_seconds() / 60;
    let remaining_minutes = MIN_INTERVAL_MINUTES - elapsed_minutes;

    if remaining_minutes <= 0 {
        Ok((true, 0))
    } else {
        Ok((false, remaining_minutes))
    }
}

/// Atomically claim a request slot for today.
/// Returns `true` if the slot was claimed (caller may proceed with the API call).
/// Returns `false` if the rate limit forbade it (count maxed or interval not elapsed).
///
/// The single UPDATE statement performs check + increment atomically, so two
/// concurrent callers cannot both succeed.
pub async fn try_record_api_request(pool: &SqlitePool) -> sqlx::Result<bool> {
    ensure_stats_for_today(pool).await?;

    let today = jst_today();
    let now_str = jst_now().format("%Y-%m-%d %H:%M:%S").to_string();

    // The WHERE clause enforces both rate-limit conditions inside the UPDATE.
    // SQLite's `(julianday(a) - julianday(b)) * 1440` gives elapsed minutes.
    let result = sqlx::query(
        r#"
        UPDATE api_request_stats
        SET last_request_time = ?,
            request_count     = request_count + 1
        WHERE reset_date = ?
          AND request_count < ?
          AND (last_request_time IS NULL
               OR (julianday(?) - julianday(last_request_time)) * 1440 >= ?)
        "#,
    )
    .bind(&now_str)
    .bind(today)
    .bind(MAX_REQUESTS_PER_DAY)
    .bind(&now_str)
    .bind(MIN_INTERVAL_MINUTES)
    .execute(pool)
    .await?;

    Ok(result.rows_affected() > 0)
}

pub async fn requests_remaining(pool: &SqlitePool) -> sqlx::Result<i64> {
    let stats = get_stats(pool).await?;
    Ok((MAX_REQUESTS_PER_DAY - stats.request_count).max(0))
}
